crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -156,26 +164,100 @@ where
     }
 
     fn headers_range(&self, range: impl RangeBounds<BlockNumber>) -> ProviderResult<Vec<Header>> {
-        self.database.headers_range(range)
+        let mut headers = Vec::new();
+        let start = match range.start_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n + 1,
+            Bound::Unbounded => 0,
+        };
+        let end = match range.end_bound() {
+            Bound::Included(&n) => n,
+            Bound::Excluded(&n) => n - 1,
+            Bound::Unbounded => self.canonical_in_memory_state.get_canonical_block_number(),
+        };
+
+        for num in start..=end {
+            if let Some(block_state) = self.canonical_in_memory_state.state_by_number(num) {
+                headers.push(block_state.block().block().header.header().clone());
```

Comment:
this is fine for now, but I suppose there's a low chance of a race condition because there can be an update in between these calls, we can track this in new issues though.